federatedscope/core/data/base_data.py

```diff
@@ -0,0 +1,173 @@
+import logging
+from federatedscope.core.data.utils import merge_data
+from federatedscope.core.auxiliaries.dataloader_builder import get_dataloader
+
+logger = logging.getLogger(__name__)
+
+
+class StandaloneDataDict(dict):
+    """
+        `StandaloneDataDict` maintain several `ClientData`.
+    """
+    client_cfgs = None
+
+    def __init__(self, datadict, global_cfg):
+        """
+
+        Args:
+            datadict: `Dict` with `client_id` as key,  `ClientData` as value.
+            global_cfg: global CfgNode
+        """
+        self.cfg = global_cfg
+        datadict = self.preprocess(datadict)
+        super(StandaloneDataDict, self).__init__(datadict)
+
+    def resetup(self, global_cfg, client_cfgs=None):
+        """
+        Resetup new configs for `ClientData`, which might be used in HPO.
+
+        Args:
+            global_cfg: enable new config for `ClientData`
+            client_cfg: enable new client-specific config for `ClientData`
+        """
+        self.cfg, self.client_cfgs = global_cfg, client_cfgs
+        for client_id, client_data in self.items():
+            if isinstance(client_data, ClientData):
+                if client_cfgs is not None:
+                    client_cfg = global_cfg.clone()
+                    client_cfg.merge_from_other_cfg(
+                        client_cfgs.get(f'client_{client_id}'))
+                else:
+                    client_cfg = global_cfg
+                client_data.setup(client_cfg)
+            else:
+                logger.warning('`client_data` is not subclass of '
+                               '`ClientData`, and cannot re-setup '
+                               'DataLoader with new configs.')
+
+    def preprocess(self, datadict):
+        """
+        Preprocess for StandaloneDataDict.
```

Comment:
The function name has told that this is about preprocessing. We'd better use one or two sentences in docstring to describe what this function actually does.